integration_test/cases/repo.exs

```diff
@@ -999,4 +999,163 @@ defmodule Ecto.Integration.RepoTest do
     Process.put(:on_log, fn _ -> flunk("logged") end)
     TestRepo.insert!(%Post{title: "1"}, [log: false])
   end
+
+  @tag :upsert
+  test "upsert with struct uses primary key as conflict target" do
+    {:ok, inserted} = TestRepo.upsert(%Post{title: "first"})
+    {:ok, _updated} = TestRepo.upsert(%Post{id: inserted.id, title: "second"})
+
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert with struct uses given columns for conflict" do
+    uuid = Ecto.UUID.generate
+
+    {:ok, _} = TestRepo.upsert(%Post{title: "first", uuid: uuid}, conflict_target: [:uuid])
+    {:ok, _} = TestRepo.upsert(%Post{title: "second", uuid: uuid}, conflict_target: [:uuid])
+
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert with struct ignored on conflict" do
+    uuid = Ecto.UUID.generate
+
+    {:ok, _} = TestRepo.upsert(%Post{title: "first", uuid: uuid},
+      conflict_target: [:uuid], on_conflict: :ignore)
+    {:ok, _} = TestRepo.upsert(%Post{title: "second", uuid: uuid},
+      conflict_target: [:uuid], on_conflict: :ignore)
+
+    post = TestRepo.one!(Post)
+    assert post.title == "first"
+  end
+
+  @tag :upsert
+  test "upsert with struct increments value on conflict" do
+    uuid = Ecto.UUID.generate
+
+    {:ok, _} = TestRepo.upsert(%Post{title: "first", uuid: uuid, visits: 0},
+      conflict_target: [:uuid], on_conflict: [inc: [visits: 1]])
+    {:ok, _} = TestRepo.upsert(%Post{title: "second", uuid: uuid, visits: 0},
+      conflict_target: [:uuid], on_conflict: [inc: [visits: 1]])
+
+    post = TestRepo.one!(Post)
+    assert post.title == "first"
+    assert post.visits == 1
+  end
+
+  @tag :upsert
+  test "upsert with struct using composite primary key" do
+    {:ok, _} = TestRepo.upsert(%CompositePk{a: 1, b: 2, name: "first"})
+    {:ok, _} = TestRepo.upsert(%CompositePk{a: 1, b: 2, name: "second"})
+
+    record = TestRepo.one!(CompositePk)
+    assert record.name == "second"
+  end
+
+  @tag :upsert
+  test "upsert with changeset uses primary key if given" do
+    insert =
+      Ecto.Changeset.cast(%Post{}, %{"title" => "first"}, ~w(title))
+      |> Ecto.Changeset.unique_constraint(:uuid)
+
+    {:ok, inserted} = TestRepo.upsert(insert)
+
+    # Ignore the UUID unique constraint when the primary key has been set
+    changes = %{"id" => inserted.id, "title" => "second"}
+    update =
+      Ecto.Changeset.cast(%Post{}, changes, ~w(id title))
+      |> Ecto.Changeset.unique_constraint(:uuid)
+    {:ok, _} = TestRepo.upsert(update)
+
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert with changeset uses unique constraint if missing primary key" do
+    insert =
```

Comment:
I agree with this. We should default to all primary keys or be explicit. Trying to guess will be too much, specially because the constraints in the changeset is for error reporting and not for handling conflict targets. In fact, if I specify a constraint in the changeset today, it is because I expect it to fail.